l2geth/rollup/sync_service.go

```diff
@@ -630,100 +509,179 @@ func (s *SyncService) SetLatestVerifiedIndex(index *uint64) {
+	// Note that Ethereum Layer one consensus rules dictate that the timestamp
+	// must be strictly increasing between blocks, so no need to check both the
+	// timestamp and the blocknumber.
+	if tx.L1Timestamp() == 0 {
+		ts := s.GetLatestL1Timestamp()
+		bn := s.GetLatestL1BlockNumber()
+		tx.SetL1Timestamp(ts)
+		tx.SetL1BlockNumber(bn)
+	} else if tx.L1Timestamp() > s.GetLatestL1Timestamp() {
+		// If the timestamp of the transaction is greater than the sync
+		// service's locally maintained timestamp, update the timestamp and
+		// blocknumber to equal that of the transaction's. This should happen
+		// with `enqueue` transactions.
+		ts := tx.L1Timestamp()
+		bn := tx.L1BlockNumber()
+		s.SetLatestL1Timestamp(ts)
+		s.SetLatestL1BlockNumber(bn.Uint64())
+		log.Debug("Updating OVM context based on new transaction", "timestamp", ts, "blocknumber", bn.Uint64(), "queue-origin", tx.QueueOrigin().Uint64())
+	} else if tx.L1Timestamp() < s.GetLatestL1Timestamp() {
+		log.Error("Timestamp monotonicity violation", "hash", tx.Hash().Hex())
+	}
+
+	if tx.GetMeta().Index == nil {
+		index := s.GetLatestIndex()
+		if index == nil {
+			tx.SetIndex(0)
+		} else {
+			tx.SetIndex(*index + 1)
+		}
+	}
+	s.SetLatestIndex(tx.GetMeta().Index)
+	if tx.GetMeta().QueueIndex != nil {
+		s.SetLatestEnqueueIndex(tx.GetMeta().QueueIndex)
+	}
+	// The index was set above so it is safe to dereference
+	log.Debug("Applying transaction to tip", "index", *tx.GetMeta().Index, "hash", tx.Hash().Hex())
+
 	txs := types.Transactions{tx}
 	s.txFeed.Send(core.NewTxsEvent{Txs: txs})
+	// Block until the transaction has been added to the chain
+	log.Trace("Waiting for transaction to be added to chain", "hash", tx.Hash().Hex())
+	<-s.chainHeadCh
+
+	return nil
+}
+
+// applyBatchedTransaction applies transactions that were batched to layer one.
+// The sequencer checks for batches over time to make sure that it does not
+// deviate from the L1 state and this is the main method of transaction
+// ingestion for the verifier.
+func (s *SyncService) applyBatchedTransaction(tx *types.Transaction) error {
+	if tx == nil {
+		return errors.New("nil transaction passed into applyBatchedTransaction")
+	}
+	index := tx.GetMeta().Index
+	if index == nil {
+		return errors.New("No index found on transaction")
+	}
+	log.Trace("Applying batched transaction", "index", *index)
+	err := s.applyIndexedTransaction(tx)
+	if err != nil {
+		return fmt.Errorf("Cannot apply batched transaction: %w", err)
+	}
+	s.SetLatestVerifiedIndex(index)
 	return nil
 }
 
 // Higher level API for applying transactions. Should only be called for
 // queue origin sequencer transactions, as the contracts on L1 manage the same
 // validity checks that are done here.
-func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
+func (s *SyncService) ValidateAndApplySequencerTransaction(tx *types.Transaction) error {
+	if s.verifier {
+		return errors.New("Verifier does not accept transactions out of band")
+	}
```

Comment:
There is a higher level check at the RPC layer here:
https://github.com/ethereum-optimism/optimism/blob/0d1ed6734877a502ada875d64f309411a2f2ac2d/l2geth/internal/ethapi/api.go#L1762

This codepath is only called by the RPC, it is a sanity check because if a verifier accepted a transaction it would result in chain corruption. A fraud prover that depends on the node would attempt to prove fraud when there isn't. We do not yet support reorganizations but should in the future